src/util/dapps.js

```diff
@@ -128,16 +132,49 @@ export function fetchBuiltinApps (api) {
 }
 
 export function fetchLocalApps (api) {
-  return api.parity.dappsList()
-    .then((apps) => {
-      return apps
-        .map((app) => {
-          app.type = 'local';
-          app.visible = true;
-          return app;
-        })
-        .filter((app) => app.id && !['ui'].includes(app.id));
-    })
+  const dappsPath = getLocalDappsPath();
+
+  return fsReaddir(dappsPath) // List files
+    .then(filenames => // Gather info about files
+      Promise.all(filenames.map(filename => {
+        const filePath = path.join(dappsPath, filename);
+
+        return fsStat(filePath).then(stat => ({ isDirectory: stat.isDirectory(), filePath, filename }));
+      }))
+    )
+    .then(files => // Only keep directories
+      files.filter(({ isDirectory }) => isDirectory))
+    .then(dappsFolders => // Parse manifests
+      Promise.all(dappsFolders.map(({ filePath, filename }) => {
+        const manifestPath = path.join(filePath, 'manifest.json');
+
+        return fsReadFile(manifestPath).then(r => {
+          try {
+            return { filename, manifest: JSON.parse(r) };
+          } catch (e) {
+            console.error(`Couldn't parse manifest.json for local dapp ${filePath}`, e);
+            return { filename };
+          }
+        }).catch(e => {
+          console.error(`Couldn't read manifest.json for ${filePath}`);
+          return { filename };
+        });
+      })))
+    .then(dapps =>
+      dapps.filter(({ manifest }) => manifest))
+    .then(dapps =>
+      dapps.map(({ filename, manifest: { id, localUrl, iconUrl, ...rest } }) => (
+        {
+          ...rest,
+          type: 'local',
+          // Prevent using the appId of an existing non-local dapp with already approved permissions
+          id: `LOCAL-${id}`,
+          folderName: 'XYZ',
+          visible: true,
+          localUrl: localUrl || `file://${dappsPath}/${filename}/index.html`,
```

Comment:
Maybe we could rename this field to `url ` like in [dappsBuiltin.json](https://github.com/parity-js/shell/blob/master/src/Dapps/dappsBuiltin.json#L130)

Reply:
hmm, I don't think `url` in dappsBuiltin.json is used anywhere AFAIK? (are you okay with me removing this field from dappsBuiltin.json?) Their actual URL is `(path to electron build)/dapps/(id)/` so there doesn't seem to be any relationship

However we could compute the localUrl of built-in dapps when parsing them [here](https://github.com/parity-js/shell/blob/master/src/util/dapps.js#L117) for consistency (add ``app.localUrl = `file://${builtinDappsPath}/${app.id}` ``); then we can remove [this logic](https://github.com/parity-js/shell/blob/61b286d2578c4d926be7edc262f6ee5b92f4172e/src/Dapp/dapp.js#L184-L198) from `Dapp.js` and have it be simply ``src = `${app.localUrl}?appId=${app.id} `` just like local dapps (we can decide to keep the query string or ditch it for built-in dapps; if we keep it `inject.js will be simpler). What do you think?

Reply:
Agreed with everything: compute it upstream, so that dapps.js looks simpler, and add query string for consistency.

If it can make everything simpler, then go for it.

Reply:
On a related note: maybe add a README.md in a folder that described all the fields in the manifest.json. Not urgent, can be done later, eg in the wiki .